EditorHelp: Improve enum ref resolving and add constant ref support

neikeq opened the pull request:
- Enum reference resolving will now search in the `@globalscope` (sorry for accidental tagging 😅) if no class is specified and the enum cannot be resolved in the current class.
- Added support for constant references, e.g.: [const KEY_ENTER] or [const Control.FOCUS_CLICK]. It supports enum constants (the enum name must not be included).

Also made some improvements to the classref regarding references. Some of them make use of the new addition or depend on the fixes introduced with this PR.

neikeq committed:
EditorHelp, makerst: Improve enum ref resolving and constant ref support

Enum reference resolving will now search in the @GlobalScope if no class is specified and the enum cannot be resolved in the current class.
Added support for constant references in EditorHelp, e.g.: [constant KEY_ENTER] or [constant Control.FOCUS_CLICK]. It supports enum constants (the enum name must not be included).

neikeq commented on doc/classes/BaseButton.xml:
```diff
@@ -48,7 +48,7 @@
 		</member>
 		<member name="button_mask" type="int" setter="set_button_mask" getter="get_button_mask">
 			Binary mask to choose which mouse buttons this button will respond to.
-			To allow both left-click and right-click, set this to 3, because it's BUTTON_MASK_LEFT | BUTTON_MASK_RIGHT.
```
I was surprised when I found this. We shouldn't be hard-coding the integral value. Even less when we also include the valid code next to it.

bojidar-bg commented:
The convertor to .rst should also be updated.

neikeq commented:
Didn't realize the branch name was used for filling the PR title...

neikeq commented:
I'm in the process of updating the rst converter. It appears it already supported a way to reference constants that is being used in a few places of the classref. The format is `[constant NAME]`. This is not supported in the EditorHelp and appears as raw text where used. It cannot be supported either because the tag type can only be 6 characters long. I will update the rst converter to use `const` instead and update the previous references in the classref.

neikeq commented:
Done!

akien-mga commented on doc/classes/LineEdit.xml:
```diff
@@ -24,7 +24,7 @@
 			</return>
 			<description>
-				Erases the [code]LineEdit[/code] text.
+				Erases the [LineEdit] text.
```
This was done on purpose to avoid self-referencing hyperlinks. The docs guidelines specify that when referencing Self in the docs, it should use `[code]` instead of `[Self]`.

neikeq replied:
Could we change this and move that logic to EditorHelp and makerst (to not generate an hyperlink if it's to the current class)? It's harder to parse the content of code tags.
I'm willing to do these changes and replace existing cases in the classref if approved.

akien-mga replied:
I'd be fine with that yeah. For EditorHelp we can likely keep the hyperlink as they don't really hurt, it's more for the HTML that we want to avoid overloading the file with irrelevant external links.

neikeq replied:
Done. Replaced all the `[code]CurrentClassName[/code]` occurences with `[CurrentClassName]` and made makerst output them as code tags as it was doing.

akien-mga commented:
> The format is [constant NAME]. This is not supported in the EditorHelp and appears as raw text where used. It cannot be supported either because the tag type can only be 6 characters long.

Is this 6-char limit meaningful though? If it's arbitrary, it can likely be removed, as it seems all checks are done with `startswith` anyway. I think it made sense to use `[constant NAME]` as it matches the corresponding XML tag `<constant>`, like all other tags. I won't fight over it though, it's fine too to use `[const NAME]`, just slightly less user-friendly for new documentation writers.

neikeq commented:
The 6 chars length is meaningful. In `editor_help.cpp` you can see it's using `rpad(6)` to make sure it has that exact length (resulting in `enum  ` and `const `). And then there is `substr(1, 6)`.
I haven't check but it may be trivial to increase this length if you want to allow longer identifiers like `constant`.

neikeq commented:
Renamed the `const` tag to `constant`.

neikeq committed:
ClassRef: Replace [code]CurrentClass[/code] with [CurrentClass]

Modified makerst to generate code tags for these to avoid hyperlinks to the same class.

neikeq commented:
Re-estructured the changes into two commits. That was harder than I expected it would be... 😓

reduz commented:
looks great

akien-mga commented:
Thanks!

hpvb commented:
Consider for 3.1.2
